jujugui/static/gui/src/app/assets/javascripts/jujulib/juju.js

```diff
@@ -152,11 +152,423 @@ var module = module;
+      query = query ? '?' + query : '';
+      if (extension) {
+        endpoint = endpoint + extension;
+      }
+      return this.url + this.version + '/' + endpoint + query;
+    },
+
+    /**
+      Transforms the results from a charmstore query into model objects.
+
+      @method _transformQueryResults
+      @param {Function} successCallback Called when the api request completes
+        successfully.
+      @param {Object} response Thre XHR response object.
+    */
+    _transformQueryResults: function(successCallback, response) {
+      var data = JSON.parse(response.target.responseText);
+      // If there is a single charm or bundle being requested then we need
+      // to wrap it in an array so we can use the same map code.
+      data = data.Results ? data.Results : [data];
+      var models = [];
+      data.forEach(function(entity) {
+        var entityData = this._processEntityQueryData(entity);
+        if (this.processEntity !== undefined) {
+          entityData = this.processEntity(entityData);
+        }
+        models.push(entityData);
+      }, this);
+      successCallback(models);
+    },
+
+    /**
+      Recursively converts all keys to lowercase when assigning them to the
+      supplied host object.
+
+      @method lowerCaseKeys
+      @param {Object} obj The source object with the uppercase keys.
+      @param {Object} host The host object in which the keys will be assigned.
+      @param {Integer} exclude Exclude a particular level from lowercasing when
+        recursing; uses a 0-based index, so if 0 is specified, the keys at the
+        first level of recursion will not be lowercased. If 3 is specified, the
+        keys at the fourth level of recursion will not be lowercased.
+      @return {Undefined} Does not return a value, modifies the supplied host
+        object in place.
+    */
+    _lowerCaseKeys: function(obj, host, exclude) {
+      if (!obj) {
+        return;
+      }
+      Object.keys(obj).forEach(function(key) {
+        // An exclude of 0 means "don't lowercase this level".
+        var newKey = key;
+        if (exclude !== 0) {
+          newKey = key.toLowerCase();
+        }
+        // Create shallow copies of objects for the copy.
+        var copy;
+        // XXX jcsackett 2015-11-09: Previously we used Y.merge to make shallow
+        // copies. As a side effect, arrays were munged into weird objects.
+        // Unfortunately the GUI relies on this side effect. When we can update
+        // the gui to not expect this behavior we can just use the statement in
+        // the else block of this if statement.
+        if (Array.isArray(obj[key])) {
+          copy = {};
+          obj[key].forEach(function(val ,i) {
+            copy[i] = val;
+          });
+        } else {
+          copy =
+              (typeof obj[key] === 'object' && obj[key] !== null) ?
+                  JSON.parse(JSON.stringify(obj[key])) : obj[key];
+        }
+        host[newKey] = copy;
```

Comment:
The above copy code replaces a use of Y.merge, as described in the XXX comment above.